addons/purchase/models/purchase.py

```diff
@@ -916,27 +916,31 @@ def _run_buy(self, product_id, product_qty, product_uom, location_id, name, orig
         for line in po.order_line:
             if line.product_id == product_id and line.product_uom == product_id.uom_po_id:
                 if line._merge_in_existing_line(product_id, product_qty, product_uom, location_id, name, origin, values):
-                    procurement_uom_po_qty = product_uom._compute_quantity(product_qty, product_id.uom_po_id)
-                    seller = product_id._select_seller(
-                        partner_id=partner,
-                        quantity=line.product_qty + procurement_uom_po_qty,
-                        date=po.date_order and po.date_order[:10],
-                        uom_id=product_id.uom_po_id)
-
-                    price_unit = self.env['account.tax']._fix_tax_included_price_company(seller.price, line.product_id.supplier_taxes_id, line.taxes_id, values['company_id']) if seller else 0.0
-                    if price_unit and seller and po.currency_id and seller.currency_id != po.currency_id:
-                        price_unit = seller.currency_id.compute(price_unit, po.currency_id)
-
-                    po_line = line.write({
-                        'product_qty': line.product_qty + procurement_uom_po_qty,
-                        'price_unit': price_unit,
-                        'move_dest_ids': [(4, x.id) for x in values.get('move_dest_ids', [])]
-                    })
+                    vals = self._update_po_line(product_id, product_qty, product_uom, values, partner, line)
+                    po_line = line.write(vals)
                     break
         if not po_line:
             vals = self._prepare_purchase_order_line(product_id, product_qty, product_uom, values, po, supplier)
             self.env['purchase.order.line'].create(vals)
 
+    def _update_po_line(self, product_id, product_qty, product_uom, values, partner, line):
+        procurement_uom_po_qty = product_uom._compute_quantity(product_qty, product_id.uom_po_id)
+        seller = product_id._select_seller(
+            partner_id=partner,
+            quantity=line.product_qty + procurement_uom_po_qty,
+            date=po.date_order and po.date_order[:10],
+            uom_id=product_id.uom_po_id)
+
+        price_unit = self.env['account.tax']._fix_tax_included_price_company(seller.price, line.product_id.supplier_taxes_id, line.taxes_id, values['company_id']) if seller else 0.0
+        if price_unit and seller and po.currency_id and seller.currency_id != po.currency_id:
+            price_unit = seller.currency_id.compute(price_unit, po.currency_id)
+
+        return {
+            'product_qty': product_qty,
```

Comment:
This is incorrect, it should be `'product_qty': line.product_qty + procurement_uom_po_qty,`